hare/eligibility/oracle.go

```diff
@@ -40,9 +40,8 @@ const (
 )
 
 const (
-	activesCacheSize = 5          // we don't expect to handle more than two layers concurrently
-	maxSupportedN    = 1073741824 // higher values result in an overflow
-	// TODO(mafa): why (MaxUint32 + 1)/4?
+	activesCacheSize = 5                       // we don't expect to handle more than two layers concurrently
+	maxSupportedN    = (math.MaxInt32 + 1) / 2 // higher values result in an overflow
```

Comment:
I traced n a bit through the code, besides logging it is only used in 2 places:

https://github.com/spacemeshos/go-spacemesh/blob/23ba30c8cecbd13059bccd1a83dfb5750921a5a1/hare/eligibility/oracle.go#L297-L300

here if `n > 2^16` the `if` will evaluate to `false`; so a max of `2^30 - 1` or `2^32 - 1` makes no difference also in terms of evaluation (if `n >= x` BinCDF will return `1` immediately).

https://github.com/spacemeshos/go-spacemesh/blob/23ba30c8cecbd13059bccd1a83dfb5750921a5a1/hare/eligibility/oracle.go#L356-L360

This is the only other place. Here `x` and `n` the moment can be up to `2^30 - 1` and are cast to an `uint16` before being returned - possibly overflowing. The returned value is only ever checked against zero, so its not a big deal but in rare cases someone who should be eligible will be randomly counted as not eligibly because they have an exact multiple of `2 ^ 16` for their `n`.

Reply:
The problem I see is that any value above 65536 for `n` will cause overflows during eligibility calculation. `n` is the miner weight (sometimes multiplied by committee size), but miner weights are defined as `uint64`s.

With a large enough weight it will overflow and transmit a `hare.Message` with `Eligibility.Count` lower than it should be.